flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/typeutils/FieldInfoUtils.java

```diff
@@ -0,0 +1,494 @@
+		if (fieldInfos.stream().anyMatch(info -> info.getFieldName().equals("*"))) {
+			throw new TableException("Field name can not be '*'.");
+		}
+
+		String[] fieldNames = fieldInfos.stream().map(FieldInfo::getFieldName).toArray(String[]::new);
+		int[] fieldIndices = fieldInfos.stream().mapToInt(FieldInfo::getIndex).toArray();
+		return new FieldsInfo(fieldNames, fieldIndices);
+	}
+
+	/**
+	 * Returns field names for a given {@link TypeInformation}.
+	 *
+	 * @param inputType The TypeInformation extract the field names.
+	 * @param <A> The type of the TypeInformation.
+	 * @return An array holding the field names
+	 */
+	public static <A> String[] getFieldNames(TypeInformation<A> inputType) {
+		validateType(inputType);
+
+		final String[] fieldNames;
+		if (inputType instanceof CompositeType) {
+			fieldNames = ((CompositeType<A>) inputType).getFieldNames();
+		} else {
+			fieldNames = new String[]{"f0"};
+		}
+
+		if (Arrays.asList(fieldNames).contains("*")) {
+			throw new TableException("Field name can not be '*'.");
+		}
+
+		return fieldNames;
+	}
+
+	/**
+	 * Validate if class represented by the typeInfo is static and globally accessible.
+	 *
+	 * @param typeInfo type to check
+	 * @throws TableException if type does not meet these criteria
+	 */
+	public static <A> void validateType(TypeInformation<A> typeInfo) {
+		Class<A> clazz = typeInfo.getTypeClass();
+		if ((clazz.isMemberClass() && !Modifier.isStatic(clazz.getModifiers())) ||
+			!Modifier.isPublic(clazz.getModifiers()) ||
+			clazz.getCanonicalName() == null) {
+			throw new TableException(format(
+				"Class '%s' described in type information '%s' must be " +
+				"static and globally accessible.", clazz, typeInfo));
+		}
+	}
+
+	/**
+	 * Returns field indexes for a given {@link TypeInformation}.
+	 *
+	 * @param inputType The TypeInformation extract the field positions from.
+	 * @return An array holding the field positions
+	 */
+	public static int[] getFieldIndices(TypeInformation<?> inputType) {
+		return IntStream.range(0, getFieldNames(inputType).length).toArray();
+	}
+
+	/**
+	 * Returns field types for a given {@link TypeInformation}.
+	 *
+	 * @param inputType The TypeInformation to extract field types from.
+	 * @return An array holding the field types.
+	 */
+	public static TypeInformation<?>[] getFieldTypes(TypeInformation<?> inputType) {
+		validateType(inputType);
+
+		final TypeInformation<?>[] fieldTypes;
+		if (inputType instanceof CompositeType) {
+			int arity = inputType.getArity();
+			CompositeType ct = (CompositeType) inputType;
+			fieldTypes = IntStream.range(0, arity).mapToObj(ct::getTypeAt).toArray(TypeInformation[]::new);
+		} else {
+			fieldTypes = new TypeInformation[]{inputType};
+		}
+
+		return fieldTypes;
+	}
+
+	public static <T> TableSchema calculateTableSchema(
+		TypeInformation<T> typeInfo,
+		int[] fieldIndexes,
+		String[] fieldNames) {
+
+		if (fieldIndexes.length != fieldNames.length) {
+			throw new TableException(String.format(
```

Comment:
a lot of this logic is also present in `TableSchema` but I guess it is ok to have it here, because the code will be removed there shortly.